replicateexistingcluster/replicateexistingcluster.go

```diff
@@ -0,0 +1,185 @@
+type Service struct {
+	existingClusterExportService *export.Service
+	simulatorExportService       *export.Service
+}
+
+// NewReplicateExistingClusterService initializes Service.
+func NewReplicateExistingClusterService(exportService *export.Service, existingClusterExportService *export.Service) *Service {
+	return &Service{
+		existingClusterExportService: existingClusterExportService,
+		simulatorExportService:       exportService,
+	}
+}
+
+// ImportFromExistingCluster get resources from existing cluster via existingClusterExportService
+// and then apply those to the simulator without restarting it via simulatorExportService.
+func (s *Service) ImportFromExistingCluster() error {
+	ctx := context.Background()
+	expRes, err := s.existingClusterExportService.Export(ctx)
+	if err != nil {
+		return xerrors.Errorf("call Export of existingClusterExportService: %w", err)
+	}
+	impRes, err := cnvtToResourcesForImportFromResourcesForExport(expRes)
+	if err != nil {
+		return xerrors.Errorf("call cnvtToResourcesForImportFromResourcesForExport: %w", err)
+	}
+	// Import to the simulator without restarting it.
+	if err := s.simulatorExportService.Import(ctx, impRes, s.simulatorExportService.IgnoreErr(), s.simulatorExportService.IgnoreRestart()); err != nil {
+		return xerrors.Errorf("call Import: %w", err)
+	}
+	return nil
+}
+
+func cnvtToResourcesForImportFromResourcesForExport(expRes *export.ResourcesForExport) (*export.ResourcesForImport, error) {
+	pods, err := cnvtPodListToApplyConfigurationList(expRes.Pods)
+	if err != nil {
+		return nil, xerrors.Errorf("call cnvtPodListToApplyConfigurationList: %w", err)
+	}
+	nodes, err := cnvtNodeListToApplyConfigurationList(expRes.Nodes)
+	if err != nil {
+		return nil, xerrors.Errorf("call cnvtNodeListToApplyConfigurationList: %w", err)
+	}
+	pvs, err := cnvtPvListToApplyConfigurationList(expRes.Pvs)
+	if err != nil {
+		return nil, xerrors.Errorf("call cnvtPvListToApplyConfigurationList: %w", err)
+	}
+	pvcs, err := cnvtPvcListToApplyConfigurationList(expRes.Pvcs)
+	if err != nil {
+		return nil, xerrors.Errorf("call cnvtPvcListToApplyConfigurationList: %w", err)
+	}
+	scs, err := cnvtStorageClassesListToApplyConfigurationList(expRes.StorageClasses)
+	if err != nil {
+		return nil, xerrors.Errorf("call cnvtStorageClassesListToApplyConfigurationList: %w", err)
+	}
+	pcs, err := cnvtPriorityClassesListToApplyConfigurationList(expRes.PriorityClasses)
+	if err != nil {
+		return nil, xerrors.Errorf("call cnvtPriorityClassesListToApplyConfigurationList: %w", err)
+	}
+	return &export.ResourcesForImport{
+		Pods:            pods,
+		Nodes:           nodes,
+		Pvs:             pvs,
+		Pvcs:            pvcs,
+		StorageClasses:  scs,
+		PriorityClasses: pcs,
+		// existingClusterExportService can't export the SchedulerConfig.
+		SchedulerConfig: nil,
+	}, nil
+}
+
+func cnvtPodListToApplyConfigurationList(pods []corev1.Pod) ([]v1.PodApplyConfiguration, error) {
```

Comment:
Is this good for this converting logic?
Actually, I want to convert a method like this.
https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/2802ed2a60aa6c168e21c44f42ffcd52a7ac05a2/scheduler/scheduler.go#L169

But I can't find it. Please give me some good idea if you have?

Reply:
Yes, it's a good idea to use a generic conversion functions to convert internal object to applyConfiguration object.

Maybe you can refer to this extractor code:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/pod-security-admission/admission/admission.go#L111-L134

In this case, it becomes:
```go
func convertToApplyConfiguration(in interface{}) (interface{}, error) {
	_in, err := json.Marshal(in)
         if err != nil {
		return nil, xerrors.Errorf("call Marshal to cnvt object: %w", err)
	}
	switch in.(type) {
	case *corev1.Pod:
		p := &applyv1.PodApplyConfiguration{}
		if err := json.Unmarshal(_in, &p); err != nil {
			return nil, xerrors.Errorf("call Marshal to cnvt Pod: %w", err)
		}
		return p, nil
        case *corev1.Node:
                ...
        case *corev1.PersistentVolume:
                ...
        case *corev1.PersistentVolumeClaim:
                ...
	default:
		return nil, xerrors.Errorf("unknown type")
	}
	return nil
}
```
@sanposhiho WDYT?

Reply:
Thank you for both advice!
I am going to implement it based on @sanposhiho 's one.
it is also good for testing, I think.